Refactor OTP to have new OTP classes to replace the OBA GTFS classes

t2gran opened the pull request:
This PR is discussed in the [opentripplanner-dev group](https://groups.google.com/forum/#!topic/opentripplanner-dev/5ij1PJAVvs4).  

Issue [2494](https://github.com/opentripplanner/OpenTripPlanner/issues/2494) describe the changes i detail.

There is a lot of commits here, feel free to squash them, but be aware that som of them is necessary to preserve the git history.

t2gran commented:
There are some broken tests. I am working on them to fix them.

novalis commented on src/main/java/org/opentripplanner/gtfs/mapping/MapCollection.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class MapCollection {
+    static <S,T> Collection<T> mapCollection(Collection<S> entites, Function<S,T> mapper) {
+        return  entites == null ? null : entites.stream().map(mapper).collect(Collectors.toList());
```
nit: "s/entites/entities/"

t2gran commented:
The "new" CalendarService is reduced to only holding data that is actually in use by OTP; hence dead code is removed. I merged the two CalendarServiceDataFactories into one, since they were identical after the removal of dead code.

fpurcell commented:
Hey @t2gran and @abyrd,

I've successfully built a graph based on this PR, which passed all of my preflight tests and seems to work pretty well.  The app is running here: http://maps7.trimet.org/ui_test/   And all of the artifacts are here: http://maps7.trimet.org/t/  

So I've got a level of confidence in this PR. From my perspective as an end user of OTP, Thomas' changes look good. 

I'm thinking to run the graph in one of our production environments to get more data points, but no promises I can do this anytime soon. I'll at least get a trickle of activity from this test instance. And if any issues come up, I'll comment here again. Bottom line, things look pretty good from here, and I  have no objections to seeing this PR pulled into master.

--Frank

barbeau commented on src/main/java/org/opentripplanner/model/Stop.java:
```diff
@@ -0,0 +1,209 @@
+ * Copyright (C) 2011 Brian Ferris <bdferris@onebusaway.org>
+ * Copyright (C) 2012 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
Could the Apache-licensed classes be pulled into a separate sub-module?  If these classes are pulled into the same module and package structure and start to get refactored/modified/etc the license is going to get really blurry, as the project is under LGPL.  I've always tried to keep a hard line between components that have different licenses.

abyrd replied:
@barbeau If by separate sub-module you mean a Maven sub-module, we went through a lot of effort to remove all sub-modules from OTP and I would be against reintroducing that structure. The alternative of putting the code in a completely separate Maven module would also make the build process much more complex because these model objects are so tightly coupled to all parts of OTP.

The license situation does deserve careful consideration but at least we know there is not a deep conflict here. OTP is under LGPL 3.0, and this OBA code is under Apache 2.0. Apache 2.0 is compatible with LGPL 3.0 (http://www.gnu.org/licenses/license-list.en.html#apache2), which is to say Apache-licensed code can be included in an LGPL project. However, as you say it starts to get messy when this code is modified. 

@bdferris is there any chance you and Google would allow us to re-license these model objects under LGPL for direct inclusion in OTP?

bdferris-v2 replied:
I can maybe take a look, but I'm guessing getting Google to sign-off on a license change is a non-trivial / painful process.

novalis commented:
How does it get messy when code is modified?

abyrd commented:
Thanks @bdferris, that's true, it sounds complicated to ask Google for a license change. Though I wasn't really asking for a license change so much as an explicit nod that it's OK to sub-license Apache code as LGPL. Actually @novalis you may understand this better than I do, and we may want to call in reinforcements from SFC on this question if we don't arrive at a clear answer. 

As I understand it, the MIT license (for example) explicitly allows sublicensing. That is to say, it gives me some rights and I can in turn pass on a subset of those rights to others. MIT is so permissive that it allows someone to simply copy MIT licensed source code into a code base that will then be released under the (L)GPL, as long as you also keep the MIT notice and authorship information. This is possible because the MIT license is a strict subset of (L)GPL. 

There also seems to be no controversy at all about Apache-GPL compatibility (though I'm not sure this compatibility falls under the technical concept of "sublicensing"). For the particular versions of these two licenses we're using, there seems to be no ambiguity about the ability to mix Apache licensed files into a GPL project as long as the composite work is distributed under the GPL.

So the only weirdness comes from the requirement that these few non-GPL files retain their Apache License headers. But when someone working on the larger LGPL project modifies that Apache-licensed file, what license are the new lines under? When someone working on the Apache licensed file deletes one line from a function, what license is that function now under? Can a third party copy that particular modified file and reuse it under the terms of the Apache License? I suppose we could declare that the atomic unit of licensing is one source file, and the entire contents of these particular files remain forever under Apache License even though they're floating in a sea of LGPL code.

In my comment to @bdferris I was just hoping we could bypass this question entirely by getting approval from the original authors to strip off the Apache headers and redistribute the code with the subset of Apache License rights provided by the LGPL.

abyrd commented:
Here's some good information in section 2.2: 
http://www.softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html

abyrd commented:
The resulting license header would look like this:

```
/* 
 This program is free software: you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public License
 as published by the Free Software Foundation, either version 3 of
 the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.

 This file incorporates work covered by the following copyright and permission notice:  

 Copyright (C) 2011 Brian Ferris <bdferris@onebusaway.org>
 Copyright (C) 2012 Google, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

abyrd commented:
That same SFLC document says: "Though the file-by-file method was widely adopted in the free software community during the past two decades, it is not an ideal method, because ensuring that notices remain correct imposes a heavy burden of individual file change tracking. We recommend that project leaders begin to reconsider the file-by-file approach, as it is error-prone and can lead to inadvertent copyright infringement and improper attribution. Any method that projects choose must be properly reviewed and maintained for accuracy. If the file-by-file method continues to serve as the canonical method for copyright inventory, we believe that substantial improvements are needed in the day-to-day care of the copyright and licensing notices in most projects."

So perhaps we could just put my draft license header (above) in a single LICENSE file and strip out the rest of them from the individual files.

barbeau commented:
>This document argues pretty clearly that "one of the most challenging licensing policies to implement is one in which permissive terms are retained on files within a larger GPL’d work." (solution 2.3). 

Yes, this is what I was referring to when I said it "gets messy".

>So perhaps we could just put my draft license header (above) in a single LICENSE file and strip out the rest of them from the individual files.

Where would this LICENSE file be stored in the project hierarchy?

On a related note, to clarify what license OTP PRs are being contributed under you may want to implement [CLAAssistant](https://github.com/cla-assistant/cla-assistant) or a similar click-through license integrated with Github PRs.  We're using CLAAssistant for OneBusAway and it's worked out well.

barbeau commented:
>If by separate sub-module you mean a Maven sub-module, we went through a lot of effort to remove all sub-modules from OTP and I would be against reintroducing that structure. The alternative of putting the code in a completely separate Maven module would also make the build process much more complex because these model objects are so tightly coupled to all parts of OTP.

@abyrd Sure, that's understandable.  I assume this means you'd also be against splitting the model objects into their own separate Git repo and Maven project.  The advantage in splitting off the model classes would be that we could more easily use the exact same model classes in OTP API clients - we had to create [our own OTP model class POJOs library project](https://github.com/CUTR-at-USF/opentripplanner-pojos) to have something we could bundle with Android clients (an aside - we also had to strip out some XML-related annotations when we first created our library, but skimming the model classes in this PR it looks like those no longer exist and the model classes would be natively-compatible on Android).  It would be great if we didn't have to maintain a separate project for the model classes used in the OTP API, but I understand the desire to keep the model classes tightly integrated in the project.

abyrd commented:
@barbeau I would be very hesitant to split anything into separate Github repos or Maven projects. In my experience this significantly complicates development and build processes, and the license situation is not problematic enough to justify that complexity.

I see what you mean about creating OTP model classes for external projects. I have not dealt with that case before and did not think of it: it is convenient to have Java model objects to deserialize API responses into if you have a Java OTP client. It is debatable though whether the internal transportation system model objects are / should be the same thing as the API model objects. @t2gran do you get the sense that your new internal OTP model objects are a separate thing from the API model objects @barbeau linked to?

abyrd commented:
@barbeau in the alternate single-license-file approach, the LICENSE file would be in the standard location at the root of the project. The license would be identical, except that it would say `this project incorporates code...` instead of `this file incorporates code...` possibly with another line to state where that code is located in the project. It makes for a lot less headers to maintain, and seems to be a perfectly acceptable alternative to per-file headers. @novalis any opinions on that approach?

barbeau commented:
>@barbeau can you also clarify what problem the CLAAssistant is intended to solve? Is it really insufficient to just state in the developer docs that OTP is blanket-released under LGPL and all contributions are also LGPL?

So I'm not a lawyer, but my understanding is that it's not sufficient to state somewhere in the project docs that contributions are under a specific license.  The contributor never explicitly agrees to this and there's no indication that they are even aware of that documentation.  We discussed this with UW's legal and they wanted a CLA for contributions to OneBusAway.

CLAAssistant solves this problem by integrating with Github in a similar way to Travis CI builds - when a PR is opened the bot checks to see if that Github user has previously agreed to a click-though CLA that the project maintainers create.  If so, it comments with a confirmation [like this](https://github.com/OneBusAway/onebusaway-android/pull/872#issuecomment-383356131).  If not, it comments asking the user to sign the CLA, [like this](https://github.com/OneBusAway/onebusaway-android/pull/867#issuecomment-378478899).  After the user clicks through to sign the CLA it will check again and repeat the above.  

CLAAssistant maintains the infrastructure for all of the above, you just need to provide a config file and CLA text via Github Gists. [Here's the config](https://gist.github.com/barbeau/3bfbb3ba20c2e5db318119da73c36747) that we use for CLAAssistant for OBA, including the CLA text itself.  This was run through UW's legal department.

OBA is under Apache, though, and OTP is under LGPL, so I'm not sure if that introduces differences in CLA requirements. I imagine that Software Freedom Conservancy needs to be in the loop for CLAs to OTP?  Maybe they have some readily available resources/boilerplate text for this?

barbeau commented:
>I did some searching around, and there is advice on this from our own organization, Software Freedom Conservancy: https://sfconservancy.org/blog/2014/jun/09/do-not-need-cla/

Great!  Nice to have a clear guidance from SFC on this.  And always nice not to have to worry about the legalese :).

t2gran commented:
I think the _internal transportation system model_ (OTP Domain Model) should be separate from the API model. Then it can be changed without breaking compatibility. It can be fine grained with rich functionality - small classes with business logic. The cost is of course the needed mapping code. The implementation in OTP today is not pure, for most classes there is an API class, but some of the OBA classes(replaced with internal model in this PR) are also exposed. 

The _internal transportation system model_ should be as close to the OTP core as possible, and not a separat maven module. I have kept the original licence agreement in the files. I also want to point out that there is _not_ a lot of functionality here - only a POJOs implementation of the GTFS specification. The only functionality left is the stuff in the CalendarServiceDataFactoryImpl. So, if necessary, I could delete all the imported classes and create them from scratch. It would probably take me 2-3 days. I have written unit tests on the mapping so the risk of doing so is small. The reason for not doing it in the first place was of cause to minimize the risk.

If we in the future switch to GraphQL, the need for a Java client API library is less. At Ruter AS we use native clients (Android, iOS) while Entur uses React Native talking to the GraphQL API. Ruter uses the [Apollo](https://github.com/apollographql/apollo-android) framework to generate the POJOs for Android, and similar for iOS and Web/JS.

drewda commented:
@opentripplanner/plc this is currently the first PR on the list for merging for 1.4, so now would be a good time to restart review and discussion.

t2gran commented:
I added a new LICENSE file and went trough all of the Java code and updated the LICENSE file and removed all Java file license headers. There are License headers in the JS files too - I did not do anything with the JS files since we might remove the GUI.

For all Java files copied into OTP from another open source project, i left a one liner at the top of the file telling which project it is taken from. I also added Copyrights and licence references for these files to the LICENSE file.

abyrd commented:
Great, thanks @t2gran for making those extensive license changes. @drewda I'll resume reviewing this one.

abyrd commented:
And in response to your message from 3 May @t2gran I agree that ideally, according to proper engineering principles, the input data model (GTFS or Netex), the internal data structures (transport network graph), and the API model should all have their own classes. However, this implies a lot of near-identical classes with code to copy between them. We need enough people and long-term commitment to maintain such things, which we seem to have now going into OTP v2. So we could indeed move toward clean separation between all these data models.

t2gran commented:
I toke a second look at the LICENSE file today an realized that I misunderstood parts of it, so I updated it according to the [article](http://www.softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html) posted by @abyrd earlier.

t2gran commented:
NOTE! DO NOT SQUASH - When merging this into OTP please keep the commits, Git history will be lost if squashed.

abyrd commented:
> On 14 Aug 2018, at 19:50, Thomas Gran <notifications@github.com> wrote:
> I toke a second look at the LICENSE file today an realized that I misunderstood parts of it, so I updated it according to the article <http://www.softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html> posted by @abyrd <https://github.com/abyrd> earlier.
> 
Looks great, yes, I think this is what we want: no license headers anywhere, one license file to cover them all, and the indentation of the other license text within the main license file really improves readability.

abyrd commented:
Another comment: we may want to restructure the package hierarchy, and/or add package-info files to explain it. The new classes are directly under org.opentripplanner.model, but we are beginning to distinguish multiple data models (transit data exchange, internal transit, web API). Since we don't have separate api model classes yet it might be premature to move things into something like `org.opentripplanner.model.internal.transit`, but we can at least add a package-info explaining our intent to eventually differentiate.

t2gran commented:
@abyrd I have read trough all your comments and I agree on most of it. When I did this I wanted to do as few changes to the original code as practical - to minimize the risk. I really, really think we must address encapsulation, performance issues and other problems - but it would be good to merge inn all the other 1.4 PRs before we change too much of the code.

I will go through all your comments shortly and comment or fix them.

abyrd commented:
A final comment on memory consumption (I don't believe the memory usage situation is a serious concern now, just something we need to understand and remember). In the mailing list message https://groups.google.com/d/msg/opentripplanner-dev/5ij1PJAVvs4/PommfsgOCQAJ you observed no difference between the pre- and post-split OTP. The code included in the message has the line that forces garbage collection at every measurement commented out. Was it commented out when the measurement CSV files were produced?

t2gran commented on src/main/java/org/opentripplanner/routing/edgetype/factory/GTFSPatternHopFactory.java:
```diff
@@ -291,15 +277,15 @@ public int compare(IndexedLineSegment a, IndexedLineSegment b) {
     private GtfsFeedId _feedId;
 
-    private GtfsRelationalDao _dao;
+    private OtpTransitService _transitService;
```
I removed all leading '_' in the code: 34736a5b5884375063b73cb655e4426dff3c0642

t2gran commented on src/main/java/org/opentripplanner/model/OtpTransitService.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Methods for accessing imported entities.
+ */
+public interface OtpTransitService {
```
´OTPTransitDAO´ to me implies the use of a database, so I fell down on ´OTPTransitService´. I can rename this to ´OTPTransitDAO´ no problem - any better name? `TransitDomainModel`?

I would like to keep the interface, because it is injected into multiple places: ´PatternHopFactory´, ´CalendarService´ and ´FareServiceFactories´ -  this make a clear contract between these components. Later I would like us to move towards a TransitDomainModel which is dobbel buffered and used by the updaters, the Raptor state generator and serialized as part of the Graph. So to make a clear separation of those modules I think an interfaces helps. But since the ´OTPTransitService´ is not part of the Graph - it might be a good idea to remove the interface for now, and introduce it later if needed.

Can we get back to this after PR #2584?

t2gran commented on src/main/java/org/opentripplanner/model/impl/OtpTransitBuilder.java:
```diff
@@ -0,0 +1,161 @@
+        generateNoneExistingIds(feedInfos);
+    }
+
+    static <T extends IdentityBean<Integer>> void generateNoneExistingIds(List<T> entities) {
```
I think several GTFS domain objects had an optional id, I removed the id from what I considered ValueObjects - making sure the ID was not used in OTP. The only one left is FeedInfo. I am not sure if it is the entire object or the id that is optional, but the OBA library generated these IDs and it might be that the only place where this break now are the tests witch do not use OBS to load data. It should be possible to get rid of this. I did not remove this because I do not know if this would break anyones GTFS import.

t2gran commented on src/main/java/org/opentripplanner/gtfs/mapping/OtpTransitDaoMapper.java:
```diff
@@ -0,0 +1,63 @@
+import org.opentripplanner.model.impl.OtpTransitBuilder;
+import org.opentripplanner.model.OtpTransitService;
+
+public class OtpTransitDaoMapper {
```
I renamed this to `GTFSToOtpTransitServiceMapper` and added some JavaDoc:
```
/**
 * This class is responsible for mapping between GTFS DAO objects and into OTP Transit model.
 * General mapping code or reusable bussiness logic should be moved into the Builder; hence
 * reusable for other import modules.
 */
```
I also added some JavaDoc to the other mappers, but these are really simple so its hard to find something useful to write - that does not state the obviuos. 
  
The Mapper is only used by the `GtfsModule`, `GtfsLibrary` and `MockGtfs` - that is in the context of loading GTFS files; It is not ment for reuse - reusable stuff goes into the builder. Hence, tagging this with an interface does not make sens to me. 

The guid line says:
> As a matter of policy, all new methods, classes, and fields should include comments 
> explaining what they are for and any other pertinent information.

Could we make this:
> As a matter of policy, all new **public** classes, methods and fields should include comments 
> explaining what they are for and any other pertinent information.

+1 for enforcing the guideline. Maybe we should include something about uint testing in the guideline, too.

t2gran commented on src/main/java/org/opentripplanner/gtfs/mapping/OtpTransitDaoMapper.java:
```diff
@@ -0,0 +1,63 @@
+    private OtpTransitService map(org.onebusaway.gtfs.services.GtfsRelationalDao data) {
+        OtpTransitBuilder builder = new OtpTransitBuilder();
+
+        builder.getAgencies().addAll(agencyMapper.map(data.getAllAgencies()));
```
I can refactor the Builder to have only the needed `add(X)`, `listX() : Iterable` and so on encapsulating the OTPTransit model creation. I will make a note of this and do it as part of the #2584 PR.

abyrd replied:
Great, this would be a good first step in the direction we want to go.

t2gran commented on src/main/java/org/opentripplanner/gtfs/mapping/RouteMapper.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.HashMap;
+import java.util.Map;
+
+class RouteMapper {
```
Not all of these are entities, but I get the idea. I have used the hole scale from reflection generic mappers to this; these are very simple and easy to read - inline with level of abstractions in OTP. @abyrd Do you want me to fix this on top of PR #2584?

abyrd replied:
I'm not advocating anything with reflection, I like the simple style. I just thought it might be nice to make it one notch more formal and give them a shallow hierarchy so instead of only the class names implying they have the same function, this shared purpose is reflected in a language construct (class hierarchy). If it's not too much trouble and we can agree that this is useful for people understanding the code in the future, then yes it would be good to include this change in #2584.

t2gran commented:
I am not sure, probably tested with and without, but the submitted log files are mote likely without. Note! The `gc()` call is just a request to do gc - the VM may or may not perform one - the most common thing is not to do a full gc. I wish I had run the test with the JVM flagg `-verbose:gc`, then it would be clear.

abyrd commented:
Thanks for the comments and updates @t2gran. What do you think about renaming `FeedId` to `FeedScopedId` before the merge?

t2gran commented:
I will do that. What should i do with the `FeedIdSerializer`, it is not needed any more. We can now annotate the `FeedScopedId` so it is serializable, but we should probably think through this and do a proper cleanup. We can do this as part of #1671 or the best is probably to split the #1671 into to Issues/PRs. One for general cleanup of the API and one for the specific RouteMatcher problem.

abyrd commented:
As a matter of procedure @t2gran can you add a line to the changelog as part of this PR?

fpurcell commented:
Hey Andrew & Thomas ... I'll get it tested before the weekend, and add an approval if it passes my test suite. 

BTW, there's quite a few files being changed, and I'm sorry I don't have the bandwidth to do anything more than build-graph, run-tests, report-back-on-results, approve-on-test-success. Thus, my approval is simply based on running a regression check. I'd like to note being slightly uncomfortable giving approval w/only a minimal scan of the actual code changes (but not uncomfortable enough to not submit a review :-)); and assume that other reviewers are comfortable with the actual code changes.

fpurcell commented:
FYI: firstly, I am able to successfully build otp.jar (using jdk 1.8.0_181). But that said, I am seeing the following:

[INFO] Compiling 875 source files to /u01/home/otp/OpenTripPlanner/target/classes
[WARNING] /u01/home/otp/OpenTripPlanner/src/main/java/org/opentripplanner/analyst/cluster/JobSimulator.java: Some input files use or override a deprecated API.
[WARNING] /u01/home/otp/OpenTripPlanner/src/main/java/org/opentripplanner/analyst/cluster/JobSimulator.java: Recompile with -Xlint:deprecation for details.
[WARNING] /u01/home/otp/OpenTripPlanner/src/main/java/org/opentripplanner/routing/graph/Graph.java: Some input files use unchecked or unsafe operations.
[WARNING] /u01/home/otp/OpenTripPlanner/src/main/java/org/opentripplanner/routing/graph/Graph.java: Recompile with -Xlint:un

...

[INFO] Running org.opentripplanner.updater.bike_rental.TestBikeRentalStationSource
Bike rental station ZAC SAINT SULPICE at 48.132100, -1.635280
Bike rental station VILLEJEAN-UNIVERSITE at 48.121075, -1.704122
Bike rental station TURMEL at 48.122475, -1.651089
Bike rental station TNB at 48.107697, -1.673695
17:42:19.994 WARN (SmooveBikeRentalDataSource.java:63) Error parsing bike rental station B08
java.lang.NumberFormatException: empty String
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1842) ~[na:1.8.0_181]
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110) ~[na:1.8.0_181]
	at java.lang.Double.parseDouble(Double.java:538) ~[na:1.8.0_181]
	at org.opentripplanner.updater.bike_rental.SmooveBikeRentalDataSource.makeStation(SmooveBikeRentalDataSource.java:59) ~[classes/:na]
	at org.opentripplanner.updater.bike_rental.GenericJsonBikeRentalDataSource.parseJSON(GenericJsonBikeRentalDataSource.java:132) [classes/:na]
	at org.opentripplanner.updater.bike_rental.GenericJsonBikeRentalDataSource.update(GenericJsonBikeRentalDataSource.java:90) [classes/:na]
	at org.opentripplanner.updater.bike_rental.TestBikeRentalStationSource.testSmoove(TestBikeRentalStationSource.java:35) [test-classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_181]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_181]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_181]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_181]
	at junit.framework.TestCase.runTest(TestCase.java:176) [junit-4.11.jar:na]
	at junit.framework.TestCase.runBare(TestCase.java:141) [junit-4.11.jar:na]
	at junit.framework.TestResult$1.protect(TestResult.java:122) [junit-4.11.jar:na]
	at junit.framework.TestResult.runProtected(TestResult.java:142) [junit-4.11.jar:na]
	at junit.framework.TestResult.run(TestResult.java:125) [junit-4.11.jar:na]
	at junit.framework.TestCase.run(TestCase.java:129) [junit-4.11.jar:na]
	at junit.framework.TestSuite.runTest(TestSuite.java:255) [junit-4.11.jar:na]
	at junit.framework.TestSuite.run(TestSuite.java:250) [junit-4.11.jar:na]
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:84) [junit-4.11.jar:na]
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365) [surefire-junit4-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273) [surefire-junit4-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238) [surefire-junit4-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159) [surefire-junit4-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:379) [surefire-booter-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340) [surefire-booter-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125) [surefire-booter-2.21.0.jar:2.21.0]
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413) [surefire-booter-2.21.0.jar:2.21.0]
Bike rental station Hamn at 60.167913, 24.952269
Bike rental station Fake at 60.000000, 24.000000
Bike rental station Foo at 61.000000, 25.000000

...

[WARNING] Tests run: 485, Failures: 0, Errors: 0, Skipped: 2

abyrd commented:
> On 23 Aug 2018, at 20:27, Frank Purcell <notifications@github.com> wrote:
> BTW, there's quite a few files being changed, and I'm sorry I don't have the bandwidth to do anything more than build-graph, run-tests, report-back-on-results, approve-on-test-success. Thus, my approval is simply based on running a regression check. I'd like to note being slightly uncomfortable giving approval w/only a minimal scan of the actual code changes (but not uncomfortable enough to not submit a review :-)); and assume that other reviewers are comfortable with the actual code changes.
> 
Hi Frank, thanks a lot for your feedback. Thomas, Gard and I have already spent quite a lot of time going over the code itself, and I believe I’ve actually read every changed file, so never fear, I requested a review from you specifically to get a different point of view. Many months back, you were the first person to test out these changes with a different data source and give the thumbs up. So the checks you performed were exactly what we were looking for.

abyrd commented:
Also, thanks for pointing out that NumberFormat warning in the tests @fpurcell. I will see if I can fix that before I merge.

abyrd commented:
@fpurcell checking out the test org.opentripplanner.updater.bike_rental.TestBikeRentalStationSource#testSmoove. This uses a JSON data source containing the following bike station definition:
```
{
         "name" : "B08 Invalid",
         "operative" : true,
         "coordinates" : "",
         "style" : "",
         "avl_bikes" : 5,
         "free_slots" : 5,
         "total_slots" : 5
      }
```
And a comment in the test states `// Invalid station without coordinates should be ignored` so this test is intentionally checking that the failure you're seeing does not entirely prevent bike data parsing.

abyrd commented:
After catching this PR up to master, we get a build error. This is due to the fact that the OBA FeedInfo objects now have String IDs instead of Integer IDs. It is easy to update the OTP code to expect Strings (I've already done this), but in the process I also realize that Agencies are also identified by a single String, rather than an AgencyAndId / FeedScopedId. This discrepancy is due to the same difference of perspective that gives us OBA's AgencyAndId versus OTP's new FeedScopedId. OTP considers entities to be unique within a feed (and all feeds must have a unique ID), OBA considers them to be unique within an Agency (and all Agencies must have a unique ID or be remapped). 

So unfortunately we'll need to somehow inject feed ID information into these IDs. But this leads me to realize that we haven't verified how this code behaves when loading multiple feeds containing the same Agency ID with colliding stop IDs (as is/was the case in New York and other places where multiple feeds defined agency 1, all with stop IDs 1, 2, 3, 4...). 

Unfortunately some additional testing is necessary here, and to avoid a regression we may need some additional code to handle feed IDs, and some additional Javadoc explaining the process by which multiple feeds are loaded and merged. It is very hard to predict the effect of some existing code, e.g. the call to `reader.setDefaultAgencyId(gtfsFeedId.getId())` in GtfsModule which seems to tell the reader to use the FeedId as the *default* AgencyId, but what about entities that have a declared AgencyId in the feed? And all the code around GTFSModule line 200 which performs mappings on the OBA side.

Now that we have our own internal model, I think we can skip these hacks and just provide the FeedId to the mapper classes, allowing them to blanket overwrite whatever OBA thinks the identifier scope string is.

We should also add automated tests that load multiple feeds with and without declared IDs, to ensure that it behaves as expected and entities do not collide. I can do that.

abyrd commented:
I've created an automated test which shows there are unexpected ID collisions (which may have existed even before these changes). The existing code is trying to make both agency IDs and feed IDs unique, and keys some internal OTP maps on Agency IDs as if they are expected to be unique. All this can be fixed with FeedScopedId now under our control. I'm going to see how the test behaves on the master branch code and propose a fix.

t2gran commented:
FeedId id changed from int to String in #2637 when switching org.onebusaway:onebusaway-gtfs from 1.3.5-conveyal-SNAPSHOT-2 to 1.3.41. @abyrd Should I fix this - I also had to do this to get the code running?

When I did this, I did my best to keep the exact same behavior - replicating both wanted and more suspicious behavior. I agree we should fix the agency id and inject feed id to it, but I think that is a separate issue - @abyrd I am not sure if you want to fix this issue before merging or as a separate issue later?

> Unfortunately some additional testing is necessary here, and to avoid a regression we may need > some additional code to handle feed IDs, and some additional Javadoc explaining the process by > which multiple feeds are loaded and merged. It is very hard to predict the effect of some existing > code, e.g. the call to reader.setDefaultAgencyId(gtfsFeedId.getId()) in GtfsModule which seems > to tell the reader to use the FeedId as the default AgencyId, but what about entities that have a 
> declared AgencyId in the feed? And all the code around GTFSModule line 200 which performs 
> mappings on the OBA side.

This should work exactly the same as before - but I have to admit that it is a long time since I did this, and I do not remember how and what tested. I do know, that I tried to add UnitTests on all new code (not the code i imported, but mapping and so on), and that I did test the feed id loading. 
@abyrd  Do you want me to write an integration test with multiple feedIds - testing the different senario you mentioned?

t2gran commented:
@abyrd I did not see you last comment - let me know if I can do some of this.

abyrd commented:
@t2gran I suspect that the "incorrect" ID collision behavior I was seeing already exists in mainline OTP. I hope you didn't go to too much effort to replicate that behavior :) What I'll do is create a separate issue for the ID collision problem, create a branch adding a new unit test in commits referencing that issue, and see how it behaves when branching off master OTP. If it fails there too, we'll just make the Integer -> String ID change on this PR and merge, and fix the collisions as a separate issue.
